### PR TITLE
RA-2171 Migrate SMS to Gov Notify

### PR DESCRIPTION
**Pull request by mrpmorris**

### CODE REVIEW

File: src/SFA.DAS.Notifications.Api.Client.UnitTests/packages.config [hunk 1]
```diff
@@ -1,10 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="Castle.Core" version="3.3.3" targetFramework="net45" />
-  <package id="Microsoft.ApplicationInsights" version="2.3.0" targetFramework="net45" />
+  <package id="Microsoft.ApplicationInsights" version="2.2.0" targetFramework="net452" />
```
Comment (dev4valtech) [inline src/SFA.DAS.Notifications.Api.Client.UnitTests/packages.config]: Higher framework version and lower package version?

File: src/SFA.DAS.Notifications.Api.UnitTests/OrchestratorsTests/NotificationOrchestratorTests/WhenSendingAnSms.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        [Test]
+        public async Task ThenItShouldSendAnSms()
+        {
+            // Act
+            await _orchestrator.SendSms(_sms);
+
+            // Assert
+            _mediator.Verify(m => m.SendAsync(It.Is<SendSmsCommand>(q => q.SystemId == _sms.SystemId)), Times.Once);
```
Comment (dev4valtech) [inline src/SFA.DAS.Notifications.Api.UnitTests/OrchestratorsTests/NotificationOrchestratorTests/WhenSendingAnSms.cs]: I haven't tried, but it seems these could be merge into one verify statement?

File: src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs [hunk 3]
```diff
@@ -44,6 +56,22 @@ protected override async Task HandleCore(SendSmsCommand command)
 
             Validate(command);
 
+            if (!Guid.TryParse(command.TemplateId, out Guid templateIdGuid))
```
Comment (dev4valtech) [inline src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs]: This check seems unnecessary, your command property is of type string. If it is ambiguous then you can change the name of the property to get rid of ambiguity. IMHO if you are not expecting a guid, it shouldn't even allow it.

File: src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs [hunk 4]
```diff
@@ -16,23 +18,33 @@ namespace SFA.DAS.Notifications.Application.Commands.SendSms
     public class SendSmsCommandHandler : AsyncRequestHandler<SendSmsCommand>
     {
         [QueueName]
+#pragma warning disable IDE1006 // Naming Styles
         public string send_notifications { get; set; }
+#pragma warning restore IDE1006 // Naming Styles
```
Comment (dev4valtech) [inline src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs]: I am failing to find where this is being used.

File: src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs [hunk 5]
```diff
@@ -16,23 +18,33 @@ namespace SFA.DAS.Notifications.Application.Commands.SendSms
...
+            INotificationsRepository notificationsRepository,
+            IMessagePublisher messagePublisher,
+            ITemplateConfigurationService templateConfigurationService,
+            ILog logger)
         {
             if (notificationsRepository == null)
                 throw new ArgumentNullException(nameof(notificationsRepository));
             if (messagePublisher == null)
                 throw new ArgumentNullException(nameof(messagePublisher));
+            if (templateConfigurationService == null)
```
Comment (dev4valtech) [inline src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs]: Really?

File: src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs [hunk 6]
```diff
@@ -44,6 +56,22 @@ protected override async Task HandleCore(SendSmsCommand command)
...
             Validate(command);
 
+            if (!Guid.TryParse(command.TemplateId, out Guid templateIdGuid))
+            {
+                var templateConfiguration = await _templateConfigurationService.GetAsync();
+                string smsServiceTemplateId = templateConfiguration.SmsServiceTemplates
+                    .SingleOrDefault(x => string.Equals(command.TemplateId, x.Id, StringComparison.InvariantCultureIgnoreCase))?.ServiceId;
+                if (string.IsNullOrEmpty(smsServiceTemplateId))
+                {
+                    throw new ValidationException($"No template mapping could be found for {command.TemplateId}");
```
Comment (dev4valtech) [inline src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs]: Can this validation be moved to the Validator class?

File: src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs (hunk 5 above)
Comment (mrpmorris) [reply src/SFA.DAS.Notifications.Application/Commands/SendSms/SendSmsCommandHandler.cs]: Personally I wouldn't have any of the null checks, so I had to choose between justifying removing them, justifying including a check for my own, or justifying breaking the pattern.

**Pull request merged**